stubs/urllib3/urllib3/contrib/socks.pyi

```diff
@@ -0,0 +1,41 @@
+from collections.abc import Mapping
+from typing import Any, ClassVar
+from typing_extensions import TypedDict
+
+from ..connection import HTTPConnection, HTTPSConnection
+from ..connectionpool import HTTPConnectionPool, HTTPSConnectionPool
+from ..poolmanager import PoolManager
+
+class _TYPE_SOCKS_OPTIONS(TypedDict):
+    socks_version: int
+    proxy_host: str | None
+    proxy_port: str | None
+    username: str | None
+    password: str | None
+    rdns: bool
+
+class SOCKSConnection(HTTPConnection):
+    def __init__(self, _socks_options: _TYPE_SOCKS_OPTIONS, *args, **kwargs) -> None: ...
+    def _new_conn(self) -> Any: ...  # "socks.socksocket":
+
+class SOCKSHTTPSConnection(SOCKSConnection, HTTPSConnection): ...
+class SOCKSHTTPConnectionPool(HTTPConnectionPool): ...
+class SOCKSHTTPSConnectionPool(HTTPSConnectionPool): ...
```

Comment:
```suggestion
class SOCKSHTTPConnectionPool(HTTPConnectionPool):
    ConnectionCls: ClassVar[type[SocksConnection]]
class SOCKSHTTPSConnectionPool(HTTPSConnectionPool):
    ConnectionCls: ClassVar[type[SocksConnection]]
```

You may have to also change this line to be a `ClassVar`: https://github.com/python/typeshed/blob/df4618ef2dec1d27e363da0fc4ba8b00a2c811fe/stubs/urllib3/urllib3/connectionpool.pyi#L47

Reply:
https://github.com/urllib3/urllib3/blob/1.26.x/src/urllib3/connectionpool.py reviewed with a variety of updates to improve coverage of this file.

If you get stuck on any of the types, comparing the implementation of the 1.26.x branch (above) and the main 2.x branch seen [here](https://github.com/urllib3/urllib3/blob/main/src/urllib3/connectionpool.py) is a great way to verify types for some of the more trickily named parameters or attributes.

Reply:
Thanks! That's more than I was expecting you to do in this PR, but that's fine 🙂

Incredibly, the new stubtest failure appears as though it's due to our stdlib stubs not having `abc.ABCMeta` as the metaclass for `io.IOBase` in the stdlib. I've filed #8468 to fix it.